docs/DesignDocs/CustomAttestation.md

```diff
@@ -0,0 +1,995 @@
+     * @param[in] config_data_size The size in bytes of config_data.
+     * @retval OE_OK on success.
+     */
+    oe_result_t (*on_register)(
+        oe_attester_plugin_t* plugin_context,
+        const void* config_data,
+        size_t config_data_size);
+
+    /**
+     * The function that gets executed when a plugin is unregistered.
+     *
+     * @param[in] plugin_context A pointer to the attester plugin struct.
+     * @retval OE_OK on success.
+     */
+    oe_result_t (*on_unregister)(
+        oe_attester_plugin_t* plugin_context);
+
+    /**
+     * Generates the attestation evidence, which is defined as the data
+     * produced by the enclave. The caller may pass in custom claims, which
+     * must be attached to the evidence and then cryptographically signed.
+     *
+     * @param[in] plugin_context A pointer to the attester plugin struct.
+     * @param[in] flags Specifying default value (0) generates evidence for local
+     * attestation. Specifying OE_EVIDENCE_FLAGS_REMOTE_ATTESTATION generates
+     * evidence for remote attestation.
+     * @param[in] custom_claims The optional custom claims list.
+     * @param[in] custom_claims_length The number of custom claims.
+     * @param[in] opt_params The optional plugin-specific input parameters.
+     * @param[in] opt_params_size The size of opt_params in bytes.
+     * @param[out] evidence_buffer An output pointer that will be assigned the
+     * address of the evidence buffer.
+     * @param[out] evidence_buffer_size A pointer that points to the size of the
+     * evidence buffer in bytes.
+     * @param[out] endorsements_buffer An output pointer that will be assigned the
+     * address of the endorsements buffer.
+     * @param[out] endorsements_buffer_size A pointer that points to the size of the
+     * endorsements buffer in bytes.
+     * @retval OE_OK on success.
+     */
+    oe_result_t (*get_evidence)(
+        oe_attester_plugin_t* plugin_context,
+        uint32_t flags,
+        const oe_claim_t* custom_claims,
+        size_t custom_claims_length,
+        const void* opt_params,
+        size_t opt_params_size,
+        uint8_t** evidence_buffer,
+        size_t* evidence_buffer_size,
+        uint8_t** endorsements_buffer,
+        size_t* endorsements_buffer_size);
+
+    /**
+     * Frees the generated attestation evidence and endorsements.
+     *
+     * @param[in] plugin_context A pointer to the attester plugin struct.
+     * @param[in] evidence_buffer A pointer to the evidence buffer.
+     * @retval OE_OK on success.
+     */
+    oe_result_t (*free_evidence)(
+        oe_attester_plugin_t* plugin_context,
+        uint8_t* evidence_buffer);
+
+    /**
+     * Frees the generated attestation endorsements.
+     *
+     * @param[in] plugin_context A pointer to the attester plugin struct.
+     * @param[in] endorsements_buffer A pointer to the endorsements buffer.
+     * @retval OE_OK on success.
+     */
+    oe_result_t (*free_endorsements)(
```

Comment:
Since evidence and endorsements are both obtained from the same function (get_evidence), why do we need two separate free functions?   Is there a common case where you'd obtain them together but need to free them at different times?

Reply:
Good point. I removed `free_endorsements` and made `free_evidence` have the endorsements parameter to make it mirror `get_evidence`.